### ISSUE TITLE
illicit: offer "statically guaranteed default" API?

### ISSUE THREAD

**Issue by anp** [anp/moxie#157]
One of the main downsides for `illicit` right now is that it can't provide calling code with a static guarantee that a type will be in scope. For types that are "always in scope" by virtue of having a default value, we should allow defining a static variable that acts as a perma-reference to the type, resolving it from the local environment each time its accessed. No more runtime panics!

**Comment by rebo** [comment]
My query is would this hide logic bugs where otherwise panic would indicate a problem.

For instance if i'm trying to get TypeX from the enviroment im probably assuming it should exist from being put there previously.

If I silently get a TypeX::default() I might not detect the issue until somewhere else in my code?

Panics are not great but maybe I would prefer a panic to a silent fail?

**Comment by anp** [comment]
Yeah, I agree about the risks of defaulting to trait impls. I'm thinking something more along the lines of 

```rust
/// The current `TypeX` in this thread's environment.
static CURRENT_TYPEX: illicit::Current<TypeX> = illlicit::guarantee(TypeX::default);
```

The init closure could be any function though.

**Comment by rebo** [comment]
H yes if it was declared like that then it would be fine i think.
